Sources/Changeset.swift

```diff
@@ -116,8 +116,345 @@ public struct Changeset {
+				// as removals.
+				entry.locationsInOld.remove(closestOld)
+				entry.occurenceInNew -= 1
+				newReferences[newPosition] = .remote(closestOld)
+				oldReferences[closestOld] = .remote(newPosition)
+			} else if entry.occurenceInNew > 0 {
+				// If no old location is left, it is treated as an inserted element.
+				inserts.insert(newPosition)
+			}
+		}
+
+		// Pass 5: Mark adjacent lines as direct moves.
+		for newPosition in 0 ..< (newCount - 1) {
+			guard case let .remote(oldPosition) = newReferences[newPosition],
+			      oldPosition + 1 < oldCount,
+			      oldIdentifiers[oldPosition + 1] == newIdentifiers[newPosition + 1],
+			      case let .table(entry) = newReferences[newPosition + 1],
+			      entry.locationsInOld.contains(oldPosition + 1) else {
+				continue
+			}
+
+			newReferences[newPosition + 1] = .remote(oldPosition + 1)
+			oldReferences[oldPosition + 1] = .remote(newPosition + 1)
+			entry.occurenceInNew -= 1
+			entry.locationsInOld.remove(oldPosition + 1)
+		}
+
+		// Pass 6: Mark adjacent lines as direct moves.
+		for oldPosition in 0 ..< (oldCount - 1) {
+			guard case let .remote(newPosition) = oldReferences[oldPosition],
+			      newPosition + 1 < newCount,
+			      oldIdentifiers[oldPosition + 1] == newIdentifiers[newPosition + 1],
+			      case let .table(entry) = newReferences[newPosition + 1],
+			      entry.locationsInOld.contains(oldPosition + 1) else {
+				continue
+			}
+
+			newReferences[newPosition + 1] = .remote(oldPosition + 1)
+			oldReferences[oldPosition + 1] = .remote(newPosition + 1)
+			entry.occurenceInNew -= 1
+			entry.locationsInOld.remove(oldPosition + 1)
+		}
+
+		// Pass 7: Compute removals. Prepare removal offsets for move elimination.
+		for oldPosition in 0 ..< oldCount {
+			if case let .table(entry) = oldReferences[oldPosition], entry.occurenceInNew == 0 {
+				removals.insert(oldPosition)
+			}
+		}
+
+		// Pass 8: Compute mutations and moves.
+		var movePaths: Set<IndexPath> = []
+		var isMoveMutating: [IndexPath: Bool] = [:]
+
+		for newPosition in 0 ..< newCount {
+			guard case let .remote(oldPosition) = newReferences[newPosition] else {
+				continue
+			}
+
+			let previousIndex = previous.index(previous.startIndex, offsetBy: C.IndexDistance(oldPosition))
+			let currentIndex = current.index(current.startIndex, offsetBy: C.IndexDistance(newPosition))
+			let isMutated = !areEqual(previous[previousIndex], current[currentIndex])
+
+			if newPosition - oldPosition != 0 {
+				let indexPath: IndexPath = [oldPosition, newPosition]
+				movePaths.insert(indexPath)
+				isMoveMutating[indexPath] = isMutated
+			} else if isMutated {
+				mutations.insert(oldPosition)
+			}
+		}
+
+		// The following two passes perform move eliminations. The algorithm is
+		// conservative and care about only one contiguous block immediately following the
+		// deletion or the insertion.
+
+		// Pass 9: Eliminating removal-caused moves.
```

Comment:
A fun fact about this is that I tried to adopt the existing approach in most ObjC/Swift implementations. But when the reproducibility tests repeatedly run sufficient large random problems with it, it eventually hit the corner cases and fails the test.

Reply:
Note that the move elimination algorithm at this point is not perfect — it does not know how to eliminate moves for scattered insertions and removals. But as mentioned above, it passes the reproducibility test even at larger problem size and iteration counts, so at least given the results it is convinced to be correct.